runtime/lua/vim/diagnostic.lua

```diff
@@ -1247,6 +1292,48 @@ function M.set(namespace, bufnr, diagnostics, opts)
     diagnostic_cache[bufnr][namespace] = diagnostics
   end
 
+  once_buf_loaded(bufnr, function()
```

Comment:
This is a whole new blob of code that future developers must study. We really need to give hints where we can. In this case, the key idea seems to be (correct if wrong):


```suggestion
  -- Compute positions, set them as extmarks, and store them in
  -- diagnostic._extmark_id (used by get_logical_location). 
  once_buf_loaded(bufnr, function()
```

None of this will be obvious in the future. We need to leave hints.